test/topology_custom/test_maintenance_mode.py

```diff
@@ -0,0 +1,95 @@
+#
+# Copyright (C) 2024-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+
+from cassandra.protocol import ConfigurationException
+from cassandra.connection import UnixSocketEndPoint
+from cassandra.cluster import Cluster
+from cassandra.policies import WhiteListRoundRobinPolicy
+from test.pylib.manager_client import ManagerClient
+
+import pytest
+import logging
+import socket
+
+logger = logging.getLogger(__name__)
+
+@pytest.mark.asyncio
+async def test_maintenance_mode(manager: ManagerClient):
+    """
+    The test checks that in maintenance mode server A is not available for other nodes and for clients.
+    It is possible to connect by the maintenance socket to server A and perform local CQL operations.
+    """
+
+    server_a, server_b = await manager.server_add(), await manager.server_add()
+    workdir = await manager.server_get_workdir(server_a.server_id)
+    socket_path = workdir + "/cql.m"
+
+    cluster = Cluster([server_b.ip_addr])
+    cql = cluster.connect()
+
+    cql.execute("CREATE KEYSPACE ks WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 1}")
+    cql.execute("CREATE TABLE ks.t (k int PRIMARY KEY, v int)")
+
+    # Token ranges of the server A
+    # [(start_token, end_token)]
+    ranges = [(int(row[0]), int(row[1])) for row in cql.execute(f"""SELECT start_token, end_token, endpoint
+                                                                    FROM system.token_ring WHERE keyspace_name = 'ks'
+                                                                    AND endpoint = '{server_a.ip_addr}' ALLOW FILTERING""")]
+
+    # Insert data to the cluster and find a key that is stored on server A.
+    for i in range(256):
+        cql.execute(f"INSERT INTO ks.t (k, v) VALUES ({i}, {i})")
+
+    # [(key, token of this key)]
+    keys_with_tokens = [(int(row[0]), int(row[1])) for row in cql.execute("SELECT k, token(k) FROM ks.t")]
+    key_on_server_a = None
+
+    for key, token in keys_with_tokens:
+        for start, end in ranges:
+            if (start < end and start <= token < end) or (start > end and (token < end or start <= token)):
```

Comment:
Also, token ranges are open on the left and closed on the right. So you actually want `start < token <= end`.
Similarly in the other case.

Reply:
The other case could also use `start >= end` (so it would also work if we started the cluster with exactly one token range -- covering the entire token ring, so we'd have start = end in this case)

Reply:
> given start <= token < end, start < end is redundant

Yes, but I prefer it this way, because it's more readable. I mean something like this: `([case 1] and [condition]) or ([case 2] and [condition])`

Reply:
ok